src/Time/TimeSteppers/RungeKutta4.hpp

```diff
@@ -0,0 +1,218 @@
+ * \f}
+ *
+ * Note that in the implementation, the expression for \f$u^{n+1}\f$ is
+ * computed simultaneously with \f$v^{(4)}\f$, so that there are
+ * actually only four substeps per step.
+ */
+class RungeKutta4 : public TimeStepper::Inherit {
+ public:
+  using options = tmpl::list<>;
+  static constexpr OptionString help = {
+      "The standard fourth-order Runge-Kutta time-stepper."};
+
+  RungeKutta4() = default;
+  RungeKutta4(const RungeKutta4&) noexcept = default;
+  RungeKutta4& operator=(const RungeKutta4&) noexcept = default;
+  RungeKutta4(RungeKutta4&&) noexcept = default;
+  RungeKutta4& operator=(RungeKutta4&&) noexcept = default;
+  ~RungeKutta4() noexcept override = default;
+
+  template <typename Vars, typename DerivVars>
+  void update_u(gsl::not_null<Vars*> u,
+                gsl::not_null<History<Vars, DerivVars>*> history,
+                const TimeDelta& time_step) const noexcept;
+
+  template <typename Vars, typename DerivVars>
+  void dense_update_u(gsl::not_null<Vars*> u,
+                      const History<Vars, DerivVars>& history,
+                      double time) const noexcept;
+
+  uint64_t number_of_substeps() const noexcept override;
+
+  size_t number_of_past_steps() const noexcept override;
+
+  double stable_step() const noexcept override;
+
+  TimeStepId next_time_id(const TimeStepId& current_id,
+                          const TimeDelta& time_step) const noexcept override;
+
+  template <typename Vars, typename DerivVars>
+  bool can_change_step_size(
+      const TimeStepId& time_id,
+      const TimeSteppers::History<Vars, DerivVars>& /*history*/) const
+      noexcept {
+    return time_id.substep() == 0;
+  }
+
+  WRAPPED_PUPable_decl_template(RungeKutta4);  // NOLINT
+
+  explicit RungeKutta4(CkMigrateMessage* /*unused*/) noexcept {}
+
+  // clang-tidy: do not pass by non-const reference
+  void pup(PUP::er& p) noexcept override {  // NOLINT
+    TimeStepper::Inherit::pup(p);
+  }
+};
+
+inline bool constexpr operator==(const RungeKutta4& /*lhs*/,
+                                 const RungeKutta4& /*rhs*/) noexcept {
+  return true;
+}
+
+inline bool constexpr operator!=(const RungeKutta4& /*lhs*/,
+                                 const RungeKutta4& /*rhs*/) noexcept {
+  return false;
+}
+
+template <typename Vars, typename DerivVars>
+void RungeKutta4::update_u(
+    const gsl::not_null<Vars*> u,
+    const gsl::not_null<History<Vars, DerivVars>*> history,
+    const TimeDelta& time_step) const noexcept {
+  const size_t substep = history->size() - 1;
+  const auto& dt_vars = (history->end() - 1).derivative();
+  const auto& U0 = history->begin().value();
+
+  switch (substep) {
+    case 0: {
+      // from (17.1.3) of Numerical Recipes 3rd Edition
+      // v^(1) = u^n + dt*RHS(u^n,t^n)/2
+      // On entry V = u^n, U0 = u^n, rhs0 = RHS(u^n, t^n),
+      // time = t^n
+      *u += 0.5 * (time_step.value() * dt_vars);
```

Comment:
Remove the extra parens here (and in other similar expressions).  `double * double * DataVector` (equivalently, `(double * double) * DataVector`) is faster than `double * (double * DataVector)`.